ot/mpz-ot-core/src/ferret/receiver.rs

```diff
@@ -0,0 +1,194 @@
+    /// * `lpn_type` - The lpn type.
+    /// * `u` - The bits received from the COT ideal functionality.
+    /// * `w` - The vector received from the COT ideal functionality.
+    pub fn setup(
+        self,
+        lpn_parameters: LpnParameters,
+        lpn_type: LpnType,
+        seed: Block,
+        u: &[bool],
+        w: &[Block],
+    ) -> Result<(Receiver<state::Extension>, LpnMatrixSeed), ReceiverError> {
+        if u.len() != lpn_parameters.k || w.len() != lpn_parameters.k {
+            return Err(ReceiverError(
+                "the length of u and w should be k".to_string(),
+            ));
+        }
+
+        let lpn_encoder = LpnEncoder::<10>::new(seed, lpn_parameters.k as u32);
+
+        Ok((
+            Receiver {
+                state: state::Extension {
+                    counter: 0,
+                    lpn_parameters,
+                    lpn_encoder,
+                    lpn_type,
+                    u: u.to_vec(),
+                    w: w.to_vec(),
+                    e: Vec::default(),
+                },
+            },
+            LpnMatrixSeed { seed },
+        ))
+    }
+}
+
+impl Receiver<state::Extension> {
+    /// The prepare precedure of extension, sample error vectors and outputs information for MPCOT.
+    /// See step 3 and 4.
+    ///
+    /// # Arguments.
+    ///
+    /// * `lpn_type` - The type of LPN parameters.
+    pub fn get_mpcot_query(&mut self) -> (Vec<u32>, usize, usize) {
+        match self.state.lpn_type {
+            LpnType::Uniform => {
+                self.state.e = self.state.lpn_parameters.sample_uniform_error_vector();
+            }
+
+            LpnType::Regular => {
+                self.state.e = self.state.lpn_parameters.sample_regular_error_vector();
+            }
+        }
+        let mut alphas = Vec::with_capacity(self.state.lpn_parameters.t);
+        for (i, x) in self.state.e.iter().enumerate() {
+            if *x != Block::ZERO {
+                alphas.push(i as u32);
+            }
+        }
+        (
+            alphas,
+            self.state.lpn_parameters.t,
+            self.state.lpn_parameters.n,
+        )
+    }
+
+    /// Performs the Ferret extension.
+    /// Outputs exactly l = n - t COTs.
+    ///
+    /// See step 5 and 6.
+    ///
+    /// # Arguments.
+    ///
+    /// * `r` - The vector received from the MPCOT protocol.
+    pub fn extend(&mut self, r: &[Block]) -> Result<(Vec<bool>, Vec<Block>), ReceiverError> {
+        if r.len() != self.state.lpn_parameters.n {
+            return Err(ReceiverError("the length of r should be n".to_string()));
+        }
+
+        // Compute z = A * w + r.
+        let mut z = r.to_vec();
+        self.state.lpn_encoder.compute(&mut z, &self.state.w);
+
+        // Compute x = A * u + e.
+        let u_block = self
+            .state
+            .u
+            .iter()
+            .map(|x| {
+                if *x {
+                    bytemuck::cast(1_u128)
```

Comment:
```suggestion
                    Block::ONE
```

Please add an associated const to `Block` so we don't rely on casts in places like this